DFLL - Opened/Closed mode support

glaeqen opened the pull request.

glaeqen committed:
Gclk0 supports mode changing for DFLL

- Gclk::swap is now only available for Gclk0<One> (when it is effectively unused)
- Changed Fll enum type to marker::Dfll<T: ModeMarker>
- Introduced few types limiting Gclk knowledge about Dfll dependencies

glaeqen committed:
Introduce basic support for multimode in DFLL

glaeqen commented on hal/src/common/thumbv7em/clock/v2/gclk.rs:
```diff
@@ -359,6 +364,56 @@ where
+    }
+}
+
+impl Gclk<Gen0, marker::Dfll<marker::ClosedMode>, One> {
```
This might be a little bit of an overkill but I let it be for now.

glaeqen replied:
@bradleyharden I could avoid keeping mode in a source/marker type if I could get away with a more general specialization for `Dfll` and have a generic closure. I went this route because I wanted a way to retrieve `Pclk0` in case user changes mode from `ClosedLoop` to `OpenLoop`. The other way around it is perfectly fine, as one can just move `Pclk0` to closure. Moving out of closure is not possible.. I think.

bradleyharden commented on hal/src/common/thumbv7em/clock/v2/sources/dfll.rs:
```diff
@@ -1,103 +1,344 @@
         self.freq()
     }
 }
+
+pub mod marker {
+    use super::{GclkSourceEnum, GclkSourceType, Sealed};
+    use core::marker::PhantomData;
+
+    pub trait ModeMarker: Sealed {}
+    /// TODO
+    /// super::ClosedMode type is polluted with a generic parameter describing
+    /// reference clock. It is undesirable to have a marker type owned by a
+    /// Gclk that knows about source of its source.
+    /// This is a reason for existence of this type.
+    pub struct ClosedMode {
+        __: (),
+    }
+
+    impl Sealed for ClosedMode {}
+    impl ModeMarker for ClosedMode {}
+
+    impl ModeMarker for super::OpenMode {}
+
+    pub struct Dfll<T: ModeMarker> {
+        __: PhantomData<T>,
+    }
+
+    impl<T: ModeMarker> Sealed for Dfll<T> {}
+
+    impl<T: ModeMarker> GclkSourceType for Dfll<T> {
+        const GCLK_SRC: GclkSourceEnum = GclkSourceEnum::DFLL;
+    }
+}
```
I realized recently that we probably shouldn't reuse existing types for the implementers of `GclkSourceType`. It's too confusing to determine which type to use, and we end up creating new types with nearly identical names.

I think we should create a dedicated type-level enum. Basically just lift `GclkSourceEnum` to the type level, with dedicated types. They can be called `Dpll0Source`, `Gclk1Source`, etc. (or similar). Basically, add a consistent prefix or suffix to the type to distinguish it as a `GclkSourceType`.

Also, there's no need for this marker approach. There can only ever be one `Dfll`. A `Gclk` does not need to know or care what `Mode` it's in. We just need a single `DfllSource` to implement `GclkSourceType`.

glaeqen replied:
> Also, there's no need for this marker approach. 

I dislike the fact that now `GClk` knows about the mode `Dfll` is using but I needed it to create impl specializations for `change_mode` methods. I'll tag you in a relevant spot so you can take a look.

I agree that reusing existing types is quite messy. For protytyping sake I just created a marker type for `ClosedMode` because in contrast to `super::OpenMode`, `super::ClosedMode<T>` is polluted with a generic type.

bradleyharden commented on hal/src/common/thumbv7em/clock/v2/sources/dfll.rs:
```diff
@@ -1,103 +1,344 @@
+type Fine = u8;
+type Coarse = u8;
+
+pub trait Mode: Sealed {}
```
I would suggest `LoopMode` here, and I would rename `OpenMode` and `ClosedMode` to `OpenLoop` and `ClosedLoop`. I think those are more descriptive and clear.

glaeqen replied:
Done.

bradleyharden commented on hal/src/common/thumbv7em/clock/v2/sources/dfll.rs:
```diff
@@ -1,103 +1,344 @@
+impl<T: PclkSourceType> Mode for ClosedMode<T> {}
+
+pub struct DfllConfig<TMode: Mode> {
+    regs: Registers,
```
We should settle on consistent naming here. I think you don't like `token`?

I wouldn't be opposed to `regs` if we were making the `Registers` API public. But I don't think that's a good idea, as I've said before. To do so, I think we would have to consider a lot more possible interactions between the PAC-like API in `Registers` and the HAL API. It's also more public API surface to maintain, but I don't think people will actually use it. If someone really wants PAC-like access, I say let them use the top-level escape hatch provided with the `PacClocks` struct at the root of `clock::v2`.

If not `regs`, then I suggest `token` for a few reasons. Token implies two things: (1) that the type has no utility on its own and (2) that it can be exchanged for something that does have utility. It's just like tokens at an arcade. Do you have any other suggestions?

glaeqen replied:
No no, it was purely my negligence with regard to naming. I fully agree with your idea of tokens. Fixed.

bradleyharden replied:
One more topic to discuss, related to this.

Right now I have `Token` as a type alias for the `Registers` struct. But I don't particularly like that. That was actually my fallback option after *really* not liking my first approach.

At first, I created independent `Token` structs that took ownership of a `Registers` struct. Then the `Config` struct (or equivalent) took ownership of the `Token`. And finally, the outermost type took ownership of the `Config`. So at the top level, to access the registers, I had code like

```rust
self.config.token.regs.method()
```

It was really verbose and I hated it. I used the `Token` alias to eliminate one layer.

```rust
self.config.token.method()
```

However, it doesn't make much sense to name something `token` but use it as registers.

Maybe the solution is to create a wrapper `Token` struct that owns a `Registers` struct. But the `Config` struct unpacks the `Token` and takes out the `Registers`. Then you would get

```rust
self.config.regs.method()
```
Then, when freeing the clock, you would recreate a `Token`.

That feels better to me. Thoughts?

glaeqen replied:
Wrapper `Token` for `Registers` that gets unpacked upon construction makes sense to me.

glaeqen committed:
Add free() methods for Dfll<{OpenMode, ClosedMode}>

glaeqen committed:
Add alias to `Registers`: `DfllToken` + rename `Mode` to `LoopMode`

- Rename `Mode` to `LoopMode`
- Rename `OpenMode` to `OpenLoop`
- Rename `ClosedMode` to `ClosedLoop`
- `regs` field & param of type `Registers` refactor to `token` of type
  `DfllToken`
